actgen/change_q.py

```diff
@@ -0,0 +1,242 @@
+        similar_actions_values = np.take(difference, similar_actions, axis=-1)
+        diff_actions_values = np.take(difference, diff_actions, axis=-1)
+        assert similar_actions_values.shape == (len(difference), int(num_total_actions / num_different_actions))
+        assert diff_actions_values.shape == (len(difference), int(num_total_actions / num_different_actions))
+
+        # apply the thresh hold, ignore small changes
+        similar_actions_values = np.where(abs(similar_actions_values) > thresh, similar_actions_values, 0)
+        diff_actions_values = np.where(abs(diff_actions_values) > thresh, diff_actions_values, 0)
+
+        # check if in right direction
+        num_same_dir_similar = sum(np.mean(similar_actions_values, axis=0) > 0) - 1
+        num_diff_dir_similar = sum(np.mean(similar_actions_values, axis=0) < 0)
+        num_same_dir_diff = sum(np.mean(diff_actions_values, axis=0) > 0)
+        num_diff_dir_diff = sum(np.mean(diff_actions_values, axis=0) < 0)
+        return num_same_dir_similar, num_diff_dir_similar, num_same_dir_diff, num_diff_dir_diff
+
+
+def main(test=False):
+    # hyper parameters for plotting
+    bogy_trial = Trial()
+    num_updates = bogy_trial.params['num_update']  # number of consecutive updates
+    delta_update = bogy_trial.params['delta_update']  # add this to the original q value each iteration of the update
+    # only changes past this percentage threshold are considered a change.
+    # this percentage is the percentage of the maximum increase the q value experienced.
+    change_percentage_thresh = bogy_trial.params['change_percentage_thresh']
+    num_total_actions = int(bogy_trial.test_env.action_space.n)
+    num_different_actions = int(num_total_actions / bogy_trial.params['duplicate'])
+    out_file_path = bogy_trial.params['out_file']
+
+    # prepare to plot, open csv to write to
+    plt.figure()
+    metrics_out_file = open(out_file_path, 'w')
+    csv_writer = csv.writer(metrics_out_file)
+    csv_writer.writerow(['number of similar actions that are updated in the same direction',
+                         'number of similar actions that are updated in the different direction',
+                         'number of different actions that are updated in the same direction',
+                         'number of different actions that are updated in the different direction'])
+
+    # figure out what are the possible actions
+    actions = range(num_total_actions)
+
+    # updates for each action
+    for a in actions:
+        # set up
+        trial = Trial(test=test)
+        s = trial.test_env.reset()  # one starting state
+        original_q_values = trial.original_q_values(s)  # q(s,a) for all a in actions
+
+        base_update_value = original_q_values[a] + delta_update
+        new_values = []
+        # update for a few consecutive times
+        for i in range(num_updates):
+            update_value = base_update_value + delta_update * i
```

Comment:
I also see that you're increasing the q-value delta with each iteration. This is probably fine, although it's unlikely that we'll reach the updated value in a single optimizer step, so we don't necessarily need to change the target each iteration. Here I'd opt for whatever results in the most concise explanation in the paper, which is probably to have a constant offset and take multiple steps toward the same target.

Reply:
The delta plots suggest that we are only moving by about 0.1, anyway, despite the first q-value target being 10.0 more than the starting point. So I doubt increasing the target each time has much of an effect.

Reply:
good point, the update is indeed happening very slow. thanks for the suggestion